db/hints/sync_point_service.cc

```diff
@@ -0,0 +1,442 @@
+}
+
+void sync_point_service::register_manager(manager& mgr) {
+    auto it = std::find_if(_managers.begin(), _managers.end(), [&mgr] (const manager& other) {
+        return &mgr == &other;
+    });
+    if (it == _managers.end()) {
+        service_logger.debug("Registering hints manager: {}", (void*)&mgr);
+        _managers.push_back(mgr);
+    }
+}
+
+void sync_point_service::unregister_manager(manager& mgr) {
+    auto it = std::find_if(_managers.begin(), _managers.end(), [&mgr] (const manager& other) {
+        return &mgr == &other;
+    });
+    if (it != _managers.end()) {
+        service_logger.debug("Unregistering hints manager: {}", (void*)&mgr);
+        _managers.erase(it);
+    }
+}
+
+future<> sync_point_service::start() {
+    register_rpc_verbs();
+    _running = true;
+    return make_ready_future<>();
+}
+
+future<> sync_point_service::stop() {
+    _running = false;
+    auto abort_tasks_of_one_kind = [] (task_map& tasks) {
+        // Because tasks can delete themselves from the map when
+        // `cancel_and_finish` is called, we must copy sync point IDs to a vector
+        std::vector<sync_point_id> ids;
+        ids.reserve(tasks.size());
+        for (auto& p : tasks) {
+            ids.push_back(p.first);
+        }
+
+        return parallel_for_each(ids, [&tasks] (sync_point_id id) {
+            return tasks.at(id).cancel_and_finish();
+        });
+    };
+
+    return when_all_succeed(
+        abort_tasks_of_one_kind(_coordinator_tasks),
+        abort_tasks_of_one_kind(_follower_tasks),
+        unregister_rpc_verbs()
+    ).discard_result();
+}
+
+future<sync_point_id> sync_point_service::create_sync_point(
+        std::vector<gms::inet_address> source_nodes,
+        std::vector<gms::inet_address> target_nodes) {
+
+    return container().invoke_on(0, [source_nodes = std::move(source_nodes), target_nodes = std::move(target_nodes)] (sync_point_service& sps) {
+        sps.check_if_is_running();
+
+        if (sps._coordinator_tasks.size() >= max_coordinator_sync_points) {
+            throw std::runtime_error(format("Too many coordinator sync points; the limit is {}", max_coordinator_sync_points));
+        }
+        auto id = utils::UUID_gen::get_time_UUID();
+        // There should be no collisions with existing waiting points
+        // because get_time_UUID is supposed to generate unique IDs
+
+        auto [it, emplaced] = sps._coordinator_tasks.try_emplace(id, sps._coordinator_tasks, "coordinator", id);
+        assert(emplaced);
+        auto& task = it->second;
+        task.run([&sps, &task, id, &source_nodes, &target_nodes] (abort_source& as) mutable {
+            return with_scheduling_group(sps._task_scheduling_group, [&sps, &task, &as, id, source_nodes = std::move(source_nodes), target_nodes = std::move(target_nodes)] () mutable {
+                return sps.coordinate_hint_waiting(as, id, std::move(source_nodes), std::move(target_nodes)).finally([&task] {
```

Comment:
Why these scheduling group games? Shouldn't everything run under the maintenance group?

Reply:
`_task_scheduling_group` is the maintenance group. The `create_sync_point` function will be called by HTTP handler and I don't know which scheduling group it uses, so I put `with_scheduling_group` there just to be sure.

For the `HINT_SYNC_POINT_CREATE` verb handler, I agree - `with_scheduling_group` is redundant there because the handler will run in the scheduling. I put it there to be consistent with `create_sync_point`.

Reply:
The REST API runs under the maintenance group.

Reply:
```c++
            with_scheduling_group(maintenance_scheduling_group, [&] {
                return ctx.http_server.listen(socket_address{ip, api_port});
            }).get();
```